Líbrerias


In [0]:
import pandas as pd
from google.colab import drive
pd.set_option('display.max_rows',None)
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
import numpy as np
import matplotlib.pyplot as plt

Elegir la carpeta de Drive como directorio local

In [0]:
drive.mount('/gdrive')
%cd /gdrive

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /gdrive
/gdrive


Leer los datos 

In [0]:
datosn = pd.read_csv('/gdrive/My Drive/Competencia/DatosNegativos2.csv')
datosp = pd.read_csv('/gdrive/My Drive/Competencia/DatosPositivos2.csv')

In [0]:
datosn.shape

(13679, 1760)

In [0]:
datosp.shape

(8322, 1761)

La idea es unir las dos bases de datosn y datosp, pero para ello nos damos cuenta que la segunda base tiene una variable adicional, listaremos las variables para saber cual es. 

In [0]:
datosn.dtypes

sequence                             object
length                                int64
molecular_weight                    float64
charge                              float64
charge_density                      float64
isoelectric_point                   float64
gravy                               float64
instability_index                   float64
aromaticity                         float64
aliphatic_index                     float64
boman_index                         float64
hydrophobic_ratio                   float64
A                                   float64
R                                   float64
N                                   float64
D                                   float64
C                                   float64
E                                   float64
Q                                   float64
G                                   float64
H                                   float64
I                                   float64
L                               

In [0]:
datosp.dtypes

Unnamed: 0                            int64
sequence                             object
length                                int64
molecular_weight                    float64
charge                              float64
charge_density                      float64
isoelectric_point                   float64
gravy                               float64
instability_index                   float64
aromaticity                         float64
aliphatic_index                     float64
boman_index                         float64
hydrophobic_ratio                   float64
A                                   float64
R                                   float64
N                                   float64
D                                   float64
C                                   float64
E                                   float64
Q                                   float64
G                                   float64
H                                   float64
I                               

Al iniciar la listas de ambas bases nos damos cuenta que la diferencia es la variable llamada Unnamed:0, vamos a visualizarla.

In [0]:
datosp.head()

Unnamed: 0                   sequence  length  ...  embed_2_98  embed_2_99  class
0           0        CGGYSGGWKRKRSTSYRCG      19  ...   -0.657932    0.072015      1
1           1           GIMISLMKKLAAHIAK      16  ...   -0.274436   -0.125522      1
2           2          FKAWRWAWRMKKLAAPS      17  ...   -0.497003    0.223890      1
3           3       WGWRDIVAGIRKVAAPVLST      20  ...   -0.524435    0.252785      1
4           4  DTHFPICIFCCGCCRQSNCGMCCKT      25  ...   -1.264345   -0.040525      1

[5 rows x 1761 columns]

Nos damos cuenta que es un simple id de los datos , por tanto podemos proceder a eliminarla 

In [0]:
datosp = datosp.drop(['Unnamed: 0'],axis=1)

Procederemos ya con las bases con el mismo número de columnas a unirlas en un solo dataframe

In [0]:
datos = pd.concat([datosn,datosp], ignore_index=True)

Las variables son númericas y poseemos 1760. Por tanto procederemos a normalizar los datos. Quitaremos la variable sequence que es string.

In [0]:
secuence = datos['sequence']
clase = datos['class']

In [0]:
datos2 = datos.drop(['sequence','class'],axis=1)

In [0]:
columnas = datos2.columns.values.tolist()

In [0]:
scaler = StandardScaler()

In [0]:
datos3 = scaler.fit_transform(datos2)

In [0]:
datos4 = pd.DataFrame(data=datos3,columns=columnas)

In [0]:
datos4.head()

length  molecular_weight    charge  ...  embed_2_97  embed_2_98  embed_2_99
0 -1.514024         -1.579527  0.253299  ...    1.220006    1.233390   -0.228330
1 -1.514024         -1.509540 -0.101136  ...    0.937539    1.196791   -0.085541
2 -1.514024         -1.411265 -0.066401  ...    1.220673    0.942420   -0.367752
3 -1.514024         -1.519080 -0.455572  ...    1.331998    0.650262    0.213560
4 -1.514024         -1.448004 -0.385748  ...    1.434744    0.705446   -0.109291

[5 rows x 1758 columns]

In [0]:
datos4['secuence'] = secuence
datos4['class'] = clase

In [0]:
datos4.head()

length  molecular_weight    charge  ...  embed_2_99  secuence  class
0 -1.514024         -1.579527  0.253299  ...   -0.228330   MAATRNK      0
1 -1.514024         -1.509540 -0.101136  ...   -0.085541   MKMDGLR      0
2 -1.514024         -1.411265 -0.066401  ...   -0.367752   MNIKHYQ      0
3 -1.514024         -1.519080 -0.455572  ...    0.213560   LLWNSIP      0
4 -1.514024         -1.448004 -0.385748  ...   -0.109291   MHPHMSY      0

[5 rows x 1760 columns]

Al analizar la variable secuence, obervamos que cada una de las letras ya tienen información numérica por tanto se toma la decisión de eliminar esta variable

In [0]:
datos5 = datos4.drop(['secuence'],axis=1)

**Método de Selección Mediante Bosques Aleatorios**


In [0]:
datos6 = datos5.drop(['class'],axis=1)

In [0]:
columnas = datos6.columns.values.tolist()

In [0]:
train, test = train_test_split(datos5,test_size=0.20,random_state=42)
train_x = train[columnas]
train_y = train['class']
test_x = test[columnas]
test_y= test['class']
rf = RandomForestClassifier(random_state=42)
rf.fit(train_x,train_y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [0]:

acc = accuracy_score(test_y,rf.predict(test_x))

In [0]:
print(dict(zip(datos6.columns, abs(rf.feature_importances_).round(2))))

{'length': 0.0, 'molecular_weight': 0.0, 'charge': 0.02, 'charge_density': 0.02, 'isoelectric_point': 0.01, 'gravy': 0.0, 'instability_index': 0.0, 'aromaticity': 0.0, 'aliphatic_index': 0.0, 'boman_index': 0.0, 'hydrophobic_ratio': 0.0, 'A': 0.0, 'R': 0.0, 'N': 0.0, 'D': 0.0, 'C': 0.0, 'E': 0.0, 'Q': 0.0, 'G': 0.0, 'H': 0.0, 'I': 0.0, 'L': 0.0, 'K': 0.01, 'M': 0.02, 'F': 0.0, 'P': 0.0, 'S': 0.0, 'T': 0.0, 'W': 0.0, 'Y': 0.0, 'V': 0.0, 'AA': 0.0, 'AR': 0.0, 'AN': 0.0, 'AD': 0.0, 'AC': 0.0, 'AE': 0.0, 'AQ': 0.0, 'AG': 0.0, 'AH': 0.0, 'AI': 0.0, 'AL': 0.0, 'AK': 0.0, 'AM': 0.0, 'AF': 0.0, 'AP': 0.0, 'AS': 0.0, 'AT': 0.0, 'AW': 0.0, 'AY': 0.0, 'AV': 0.0, 'RA': 0.0, 'RR': 0.0, 'RN': 0.0, 'RD': 0.0, 'RC': 0.0, 'RE': 0.0, 'RQ': 0.0, 'RG': 0.0, 'RH': 0.0, 'RI': 0.0, 'RL': 0.0, 'RK': 0.0, 'RM': 0.0, 'RF': 0.0, 'RP': 0.0, 'RS': 0.0, 'RT': 0.0, 'RW': 0.0, 'RY': 0.0, 'RV': 0.0, 'NA': 0.0, 'NR': 0.0, 'NN': 0.0, 'ND': 0.0, 'NC': 0.0, 'NE': 0.0, 'NQ': 0.0, 'NG': 0.0, 'NH': 0.0, 'NI': 0.0, 'NL': 0.0,

In [0]:
mask = rf.feature_importances_.round(2)>0.0

In [0]:
print(mask)

[False False  True ... False False False]


In [0]:
reduced_x = datos6.loc[:, mask]

In [0]:
print(reduced_x.columns)

Index(['charge', 'charge_density', 'isoelectric_point', 'K', 'M', 'KK',
       '_SolventAccessibilityC1', '_SolventAccessibilityC3', '_ChargeC1',
       '_ChargeC3', '_ChargeT23', '_SolventAccessibilityD1001',
       '_SolventAccessibilityD3001', '_SolventAccessibilityD3100',
       '_ChargeD3001', '_ChargeD3025', '_ChargeD3075', '_ChargeD3100',
       'GearyAuto_AvFlexibility4', 'GearyAuto_FreeEnergy5', 'QSOgrant13',
       'QSOSW13'],
      dtype='object')


Las variables listadas anterior fueron las que arrojo el método de selección correctamente. 

**Método de Extracción**